src/registry/artifactory.rs

```diff
@@ -25,72 +29,99 @@ pub struct Artifactory {
     client: reqwest::Client,
 }
 
+#[derive(Error, Debug)]
+#[error("{0} is not a supported version requirement")]
+struct UnsupportedVersionRequirement(VersionReq);
+
 impl Artifactory {
-    pub fn new(registry: RegistryUri, credentials: &Credentials) -> eyre::Result<Self> {
+    /// Creates a new instance of an Artifactory registry client
+    pub fn new(registry: RegistryUri, credentials: &Credentials) -> miette::Result<Self> {
         Ok(Self {
             registry: registry.clone(),
             token: credentials.registry_tokens.get(&registry).cloned(),
             client: reqwest::Client::builder()
                 .redirect(reqwest::redirect::Policy::none())
-                .build()?,
+                .build()
+                .into_diagnostic()?,
         })
     }
 
-    /// Pings artifactory to ensure registry access is working
-    pub async fn ping(&self) -> eyre::Result<()> {
-        let repositories_url: Url = {
-            let mut uri = self.registry.to_owned();
-            let path = &format!("{}/api/repositories", uri.path());
-            uri.set_path(path);
-            uri.into()
-        };
-
-        let mut request = self.client.get(repositories_url);
+    async fn make_auth_request(
```

Comment:
I dont like this change, it creates hard ti track, oop-like patterns.

Reply:
I don't understand, all this does is move some common logic to a function so we don't need to repeat it a bunch of times.

Reply:
(Im not critical about the fact that we abstract common logic)

Reply:
I'm not sure how that would look like (I don't have much experience with pure functional languages). Since the existing code is imperative, my bias would be to keep a trivial refactor here and file an issue for changing the paradigm later, but if you have a simple change in mind happy to incorporate it (I just don't know what it would be).

Reply:
So you can usually use data lenses to extract the data you want from the parent data type, to then detect whether you want to error or not.

While writing / thinking about it a second time i found that you dont really do data introspection, but rather error validation here, so the most idiomatic thing imo would be to use the type system as much as possible:

```rust
struct Validated(Response);

impl TryFrom<Response> for Validated {
    type Error = miette::Error;

    fn try_from(/**/) -> Self {
        // do the ensuring and miette stuff
        
        Self(res)
    }
}

impl TryFrom<Validated> for Package {
// ...
}

// using in the toplevel function that does IO
fn download(&self, dep: Dependency) -> miette::Result<()> {
    let request = // build the request
    let response = response.await?;
    let package: Package = Validated::try_from(response)?.try_into()?;
    Ok(package)
}

Reply:
Thanks for the detailed explanation, Mara! This makes it all much clearer.

From what I gather, there are two points where the code looks imperative:
* The signature of `make_auth_request()`, which takes an Option argument.
  * Here I agree a builder pattern would be nicer, though I'm a little skeptical about the complexity trade-off. Still, that's something I won't argue against.
* The construction of the request, which conditionally re-binds the builder variable a couple of times.

I believe the second part looks imperative is because the `reqwest` crate chose to implement `RequestBuilder` methods consume the builder, so we are forced to either chain only once or reassign. Ideally the builder API would take optionals so we could apply the functional style unconditionally whether the token and body are present or not.

I think moving the request building logic into a separate function is an ok change to make, although it is never used in isolation it does constitute a self-contained unit of behaviour.

Ultimately, IMO the nice thing here would be to define a ValidatedRequest struct that encapsulates the underlying `reqwest` request. Then we could fix both problems at once:

```rust
impl Artifactory {
    fn new_request(&self, method: Method, url: Url) -> ValidatedRequestBuilder {
        let mut builder = ValidatedRequest::new(method, url);
        if let Some(token) = &self.token {
            builder = builder.token(token);
        }
        builder
    }

    pub fn publish(&self, package: Package, repository: String) -> miette::Result<()> {
        // ...
        let response = self.new_request(method, url).body(data).send().await?;
        // ...
    }
}
```

A few notes:
* Functions don't need to specify the `body` argument if it's not used.
* Functions never need to specify the token argument explicitly, so there's no chance of forgetting to do that.
* `ValidatedRequest::send()` takes care of common response validation internally.
* We could make `new_request()` look more functional internally by making `.token()` take an `Option<String>`, but I think that's a bad trade-off.
  * We could also make `.token()` take a mutable reference to the builder (instead of consuming it) so that variable reassignment wouldn't be necessary. Not sure if that's worth the trade-off either, but I don't feel strongly about this point.

Honestly, a lot of the problems here are due to `reqwest`'s shortcomings. If we could provide the token on client construction (rather than at request construction), we wouldn't need the `new_request()` method. If we could define common validation logic, we wouldn't need the custom `ValidatedRequest` type either.

I'll go ahead with this style, and happy to adjust as necessary, but please let's not focus too much on this as improving this code is really not the goal of this PR.

Reply:
Ok, we chatted a bit offline and with some additions (and a less confusing name for the new request type) we agreed this is best.